controller/gke-cluster-config-handler.go

```diff
@@ -0,0 +1,349 @@
+				config.Status.Phase = gkeConfigUpdatingPhase
+			}
+		}
+		config.Status.FailureMessage = message
+
+		var recordErr error
+		config, recordErr = h.gkeCC.UpdateStatus(config)
+		if recordErr != nil {
+			logrus.Errorf("Error recording gkecc [%s] failure message: %s", config.Name, recordErr.Error())
+		}
+		return config, err
+	}
+}
+
+func (h *Handler) OnGkeConfigRemoved(key string, config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+	if config.Spec.Imported {
+		logrus.Infof("cluster [%s] is imported, will not delete GKE cluster", config.Name)
+		return config, nil
+	}
+	if config.Status.Phase == gkeConfigNotCreatedPhase {
+		// The most likely context here is that the cluster already existed in GKE, so we shouldn't delete it
+		logrus.Warnf("cluster [%s] never advanced to creating status, will not delete GKE cluster", config.Name)
+		return config, nil
+	}
+
+	logrus.Infof("deleting cluster [%s]", config.Name)
+
+	//TODO: delete cluster
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	svc, err := getServiceClient(ctx, config.Spec.CredentialContent)
+	if err != nil {
+		return config, err
+	}
+
+	logrus.Debugf("Removing cluster %v from project %v, region/zone %v", config.Spec.ClusterName, config.Spec.ProjectID, config.Spec.Region)
+	operation, err := waitClusterRemoveExp(ctx, svc, config)
+
+	if err != nil && !strings.Contains(err.Error(), "notFound") {
+		return config, err
+	} else if err == nil {
+		logrus.Debugf("Cluster %v delete is called. Status Code %v", config.Spec.ClusterName, operation.HTTPStatusCode)
+	} else {
+		logrus.Debugf("Cluster %s doesn't exist", config.Spec.ClusterName)
+	}
+
+	return config, nil
+}
+
+func (h *Handler) create(config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	svc, err := getServiceClient(ctx, config.Spec.CredentialContent)
+	if err != nil {
+		return config, err
+	}
+
+	operation, err := svc.Projects.Locations.Clusters.Create(
+		locationRRN(config.Spec.ProjectID, config.Spec.Region),
+		generateClusterCreateRequest(config)).Context(ctx).Do()
+
+	if err != nil && !strings.Contains(err.Error(), "alreadyExists") {
+		return config, err
+	}
+
+	if err == nil {
+		logrus.Debugf("Cluster %s create is called for project %s and region/zone %s. Status Code %v",
+			config.Spec.ClusterName, config.Spec.ProjectID, config.Spec.Region, operation.HTTPStatusCode)
+	}
+
+	config = config.DeepCopy()
+	config.Status.Phase = gkeConfigCreatingPhase
+	return h.gkeCC.UpdateStatus(config)
+}
+
+func (h *Handler) checkAndUpdate(config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+
+	//TODO: implement -- currently a noop.
+	return config, nil
+}
+
+func (h *Handler) waitForCreationComplete(config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+	lastMsg := ""
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for {
```

Comment:
Instead of looping, you should "re-enqueue" the cluster into the operator loop. This will ensure you have the latest version of the cluster config for when you update the phase. You can see the `eksEnqueueAfter` method in the `eks-operator` for an example.

Reply:
This should be addressed.